platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/ReusableBucketPool.java

```diff
@@ -0,0 +1,158 @@
+     * Index in the array of the next empty slot. When a bucket is released to the pool,
+     * it's put to the array at this index, and the index is increased by one. If another
+     * thread is waiting for a bucket to be released at the given index, it is notified.
+     */
+    private final AtomicInteger nextE = new AtomicInteger(0);
+
+    /**
+     * Creates a new reusable bucket pool of the default size.
+     *
+     * @param serializer Key serializer used by the buckets in the pool
+     */
+    public ReusableBucketPool(final BucketSerializer<K> serializer) {
+        this(DEFAULT_POOL_SIZE, serializer);
+    }
+
+    /**
+     * Creates a new reusable bucket pool of the specified size.
+     *
+     * @param serializer Key serializer used by the buckets in the pool
+     */
+    public ReusableBucketPool(final int size, final BucketSerializer<K> serializer) {
+        poolSize = size;
+        buckets = new AtomicReferenceArray<>(poolSize);
+        locks = new AtomicReferenceArray<>(poolSize);
+        for (int i = 0; i < poolSize; i++) {
+            buckets.set(i, new Bucket<>(serializer.getKeySerializer(), this));
+            locks.set(i, new Object());
+        }
+    }
+
+    /**
+     * Gets a bucket from the pool. If the pool is empty, the calling thread waits
+     * until a bucket is released to the pool.
+     *
+     * @return A bucket that can be used for reads / writes until it's released back
+     * to the pool
+     */
+    public Bucket<K> getBucket() {
+        // Every call to this method is bound to a particular index in the pool
+        final int index = nextA.getAndUpdate(t -> (t + 1) % poolSize);
+        // Try optimistic get first
+        Bucket<K> bucket = buckets.getAndSet(index, null);
+        if (bucket == null) {
+            final Object lock = locks.get(index);
+            synchronized (lock) {
+                bucket = buckets.getAndSet(index, null);
+                // Wait until a bucket at the given index is released
+                while (bucket == null) {
+                    try {
+                        lock.wait();
+                    } catch (final InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                        throw new RuntimeException(e);
+                    }
+                    bucket = buckets.getAndSet(index, null);
+                }
+            }
+        }
+        bucket.clear();
+        return bucket;
+    }
+
+    /**
+     * Releases a bucket back to this pool. The bucket cannot be used after this
+     * call, until it's borrowed from the pool again using {@link #getBucket()}.
+     *
+     * @param bucket A bucket to release to this pool
+     */
+    public void releaseBucket(final Bucket<K> bucket) {
+        assert bucket.getBucketPool() == this;
+        int index = nextE.getAndUpdate(t -> (t + 1) % poolSize);
+        boolean released = buckets.compareAndSet(index, null, bucket);
+        while (!released) {
```

Comment:
could you please describe a scenario in which `released` is false? Is it supposed to be a really short period of time? Because if not, then "busy wait" is questionable.

Reply:
Yes, it's expected to be short period of time. I could reproduce this issue only if the size of the pool is very small, like under 100 buckets, and heavy load.

This loop is to handle scenario, when a bucket is released, the thread waiting on the corresponding index is notified, but not resumed yet and not acquired the released bucket. It may happen when so many buckets are released that `nextE` index is overflowed (i.e. more buckets are released than the size of the pool).

Reply:
Each thread calling either `getBucket` or `releaseBucket` acquires a slot in `buckets` and synchronizes on its content with `compareAndSet/getAndSet/lock`. We thus create pairs of threads working with each other: a getter will wait for a particular releaser and may remain blocked even if other buckets become available.
It's OK for now as this change removes the bottleneck from flushing but we may need to revisit this implementation in the future.